### ISSUE TITLE
Freeze, crash and low video quality in group video call (>15) in older iPhones

### ISSUE THREAD

**Issue by sindunarasimhan** [twilio/twilio-video-ios#90]
### Description

[Description of the issue]
We are building a group video call application that can support up to 25 callers concurrently for 20 minutes. Here are the following issues we are faced with in such calls. We use simulcast configurations by default. 

1. Video quality degradation after 10 minutes for small groups (<10) 
2. Low video quality for large groups  (15 - 20) 
3. Freezes and crashes in older iPhones (Specifically tested with iPhone 7S, 8 and X) 
4. Overheating in older iPhones along with quick battery drain

#### Code
       func checkVideoSenderSettings(room: Room) {
           guard let localParticipant = room.localParticipant else {
               return
           }
           guard let camera = camera else {
               return
           }

           /*
            * Update the CameraCapturer's format and the LocalParticipant's EncodingParameters based upon the size of the Room.
            * When simulcast is not used, it is preferrable for the Participant to send a smaller video stream that may be
            * easily consumed by all subscribers. If simulcast is enabled, then the source should produce frames that are
            * large enough for the encoder to create 3 spatial layers.
            *
            * A lower frame rate is used in multi-party to reduce the cumulative receiving / decoding / rendering cost for
            * subscribed video.
            */
          // let isMultiparty = room.remoteParticipants.count > 1
        
        let bitrate = CaptureDeviceUtils.kMultipartyVideoBitrate
        localParticipant.setEncodingParameters(EncodingParameters(audioBitrate: MultichatViewController.audioBitrate,
                                                                     videoBitrate: bitrate))
           
        let frontCamera = CameraSource.captureDevice(position: .front)
        let backCamera = CameraSource.captureDevice(position: .back)

        let device = frontCamera != nil ? frontCamera! : backCamera!

        let format = CaptureDeviceUtils.selectVideoFormat(multiparty: true, device:device)
            camera.selectCaptureDevice(device, format: format, completion: { (device, format, error) in
                if let error = error {
                    self.logMessage(messageText: "Failed to select format \(format), error = \(String(describing: error))")
                }
            })

       }

#### Expected Behavior
Consistent video quality without freezing, overheating etc. 

#### Actual Behavior
1. Video quality degradation after 10 minutes for small groups (<10) 
2. Low video quality for large groups  (15 - 20) 
3. Freezes and crashes in older iPhones (Specifically tested with iPhone 7S, 8 and X) 
4. Overheating in older iPhones along with quick battery drain

#### Reproduces How Often

[What percentage of the time does it reproduce?]
100% - High Frequency 

#### Logs
Debug level logs are helpful when investigating issues. To enable debug level logging, add the following code to your application:

```.swift
TwilioVideoSDK.setLogLevel(.debug)
```

```
// Log output when the issue occurs
```

### Versions

All relevant version information for the issue.

#### Video iOS SDK
3.2

#### Xcode
11.3.1

#### iOS Version
13.3.1

#### iOS Device
iPhone 7s, 8, 8+ and iPhone X

**Comment by ceaglest** [comment]
Hi @sindunarasimhan,

Thanks for writing in.

> Overheating in older iPhones along with quick battery drain

We are adding support for [Bandwidth Profiles](https://www.twilio.com/docs/video/tutorials/using-bandwidth-profile-api) on mobile to match the JavaScript SDK. This feature allows you to prioritize which tracks are most important to your app and how much bandwidth is used for receiving video. At the moment the Server does not know which sizes and frame rates your subscribers require and treats each RemoteVideoTrack equally. This is the focus of the next minor release of our iOS SDK.

> Video quality degradation after 10 minutes for small groups (<10)
> Low video quality for large groups (15 - 20)

What is the layout of your application UI? Are all of the 15-20 RemoteVideoTracks visible on screen at any given time? If they are not then you can use the [Track Subscription REST API](https://www.twilio.com/docs/video/api/track-subscriptions) to unsubscribe to any tracks that are not currently visible. This will prevent wasting resources receiving video that is not being used.

> > Freezes and crashes in older iPhones (Specifically tested with iPhone 7S, 8 and X)

Can you share a stack trace of the crash? We can look into this further.

Best,
Chris

**Comment by ceaglest** [comment]
One comment about the code snippet shared, since it looks like a modified version of MultipartyExample.

```.swift
let bitrate = CaptureDeviceUtils.kMultipartyVideoBitrate
```

If you are using simulcast this should be:

```.swift
CaptureDeviceUtils.kSimulcastVideoBitrate
```

**Comment by sindunarasimhan** [comment]
Thanks for commenting. This is helpful. I have updated the screenshot of our layout - Speaking in Swift terms, the dominant speaker changes in the main view and the collection view has thumbnails of participants - I do think selectively subscribing to the dominant speaker, and participant tracks of the collection view that is visible could be a good alternative that I can try experimenting with, but will the refresh rates be responsive enough to user scrolls or quick switching of dominant speaker? 

![img1](https://user-images.githubusercontent.com/9439749/79620762-d7767a00-80c5-11ea-9e07-800b8c912386.png)

I do have the logs, but they have some personal info like user phone number - I can email them to you personally.

**Comment by ceaglest** [comment]
Hi @sindunarasimhan,

> I do think selectively subscribing to the dominant speaker, and participant tracks of the collection view that is visible could be a good alternative that I can try experimenting with, but will the refresh rates be responsive enough to user scrolls or quick switching of dominant speaker?

Using the subscription APIs might not be ideal due to delay. Ideally, you could use a bandwidth profile to restrict the maximum number of tracks that you receive at one time and switch off the rest of the tracks that are not on screen.

We are currently working on the release supporting bandwidth profiles, and will let you know once it is ready for public use.

Best,
Chris

**Comment by piyushtank** [comment]
We have shipped Bandwidth Profile APIs in our [3.x](https://www.twilio.com/docs/video/changelog-twilio-video-ios-version-3x#bandwidth-profile) and also shipped many more features in our Video SDK. Try out our latest [4.1](https://www.twilio.com/docs/video/changelog-twilio-video-ios-latest#410-january-13-2020) SDK and let us know.

Closing the ticket as it's very old. Feel free to reopen or open a new issue if you run into any issues.
